GPU/GPUTracking/Merger/GPUTPCGMMerger.cxx

```diff
@@ -2036,35 +2046,111 @@ GPUd() void GPUTPCGMMerger::MergeLoopers(int nBlocks, int nThreads, int iBlock,
-      float d = CAMath::Sum2((params[i].x - params[j].x) * (1.f / 5.f), (params[i].y - params[j].y) * (1.f / 5.f), (params[i].tgl - params[j].tgl) * (1.f / 0.15f), dqpt / CAMath::Min(params[i].qpt, params[j].qpt) * (1.f / 0.15f));
-      //bool EQ = CAMath::Abs(params[i].x - params[j].x) < 10.f && CAMath::Abs(params[i].y - params[j].y) < 10.f && CAMath::Abs(params[i].tgl - params[j].tgl) < 0.15f && CAMath::Abs((params[i].qpt - params[j].qpt) / CAMath::Min(params[i].qpt, params[j].qpt)) < 0.15f;
-      bool EQ = d < 1.5f;
-#if GPUCA_MERGE_LOOPER_MC
+      const float d2xy = CAMath::Sum2(params[i].x - params[j].x, params[i].y - params[j].y);
+      if (d2xy > 15.f) {
+        //bs |= 1;
+        continue;
+      }
+      const auto& trk1 = mOutputTracks[params[i].id];
+      const auto& trk2 = mOutputTracks[params[j].id];
+      const auto& param1 = trk1.GetParam();
+      const auto& param2 = trk2.GetParam();
+      if (CAMath::Abs(param1.GetDzDs()) > 0.03f && CAMath::Abs(param2.GetDzDs()) > 0.03f && param1.GetDzDs() * param2.GetDzDs() * param1.GetQPt() * param2.GetQPt() < 0) {
+        //bs |= 2;
+        continue;
+      }
+
+      const float dznormalized = (CAMath::Abs(params[j].refz) - CAMath::Abs(params[i].refz)) / (CAMath::TwoPi() * 0.5f * (CAMath::Abs(param1.GetDzDs()) + CAMath::Abs(param2.GetDzDs())) * 1.f / (0.5f * (CAMath::Abs(param1.GetQPt()) + CAMath::Abs(param2.GetQPt())) * CAMath::Abs(Param().polynomialField.GetNominalBz())));
+      const float phasecorr = CAMath::Modf((CAMath::ASin(param1.GetSinPhi()) + trk1.GetAlpha() - CAMath::ASin(param2.GetSinPhi()) - trk2.GetAlpha()) / CAMath::TwoPi() + 5.5f, 1.f) - 0.5f;
+      const float phasecorrdirection = (params[j].refz * param1.GetQPt() * param1.GetDzDs()) > 0 ? 1 : -1;
+      const float dzcorr = dznormalized + phasecorr * phasecorrdirection;
+      const bool sameside = !(trk1.CSide() ^ trk2.CSide());
+      const float dzcorrlimit[4] = {sameside ? 0.018f : 0.012f, sameside ? 0.12f : 0.025f, 0.14f, 0.15f};
+      const int dzcorrcount = sameside ? 4 : 2;
+      bool dzcorrok = false;
+      float dznorm = 0.f;
+      for (int k = 0; k < dzcorrcount; k++) {
+        const float d = CAMath::Abs(dzcorr - 0.5f * k);
+        if (d <= dzcorrlimit[k]) {
+          dzcorrok = true;
+          dznorm = d / dzcorrlimit[k];
+          break;
+        }
+      }
+      if (!dzcorrok) {
+        //bs |= 4;
+        continue;
+      }
+
+      const float dtgl = param1.GetDzDs() - (param1.GetQPt() * param2.GetQPt() > 0 ? param2.GetDzDs() : -param2.GetDzDs());
+      const float dqpt = (CAMath::Abs(param1.GetQPt()) - CAMath::Abs(param2.GetQPt())) / CAMath::Min(param1.GetQPt(), param2.GetQPt());
+      float d = CAMath::Sum2(dtgl * (1.f / 0.03f), dqpt * (1.f / 0.04f)) + d2xy * (1.f / 4.f) + dznorm * (1.f / 0.3f);
+      bool EQ = d < 6.f;
```

Comment:
I understand that the code is not final yet, but is not it better to add at least most important cuts as settable parameters?

Reply:
I was thinking about that, but I didn't want to add that many more parameters. Or at least, not to this general GPU_rec object I have right now. Will add a GPU_recTPC soon that should comprise all TPC reco settings, which will contain also these.